### PR TITLE
[PyTorch] DeepCFR Implementation

### PR DESCRIPTION
**Pull request by MJ10**
The algorithms in the library are primarily written in Tensorflow. However, given the generic nature of the rest of the library, algorithms can be written with any learning library. Adding an example demonstrating how to use PyTorch would certainly be helpful for the community, as also mentioned in the [documentation](https://openspiel.readthedocs.io/en/latest/contributing.html#roadmap-and-call-for-contributions).

In this PR I have added a minimal notebook that implements DeepCFR in PyTorch which can serve as a reference.

Feedback welcome!

### CODE REVIEW

**Comment by lanctot** [comment]
This is awesome! Thanks for submitting our first PyTorch contribution! 

> The algorithms in the library are primarily written in Tensorflow. However, given the generic nature of the rest of the library, algorithms can be written with any learning library. Adding an example demonstrating how to use PyTorch would certainly be helpful for the community, as also mentioned in the [documentation](https://openspiel.readthedocs.io/en/latest/contributing.html#roadmap-and-call-for-contributions).

Agree, and in addition to TF and PyTorch, we welcome JAX contributions as well! :)

> In this PR I have added a minimal notebook that implements DeepCFR in PyTorch which can serve as a reference.
> 
> Feedback welcome!

I would love for this to be part of the main library, not just an example in a notebook. How do you feel about making it a callable module like all the other algorithms (maybe in a new subdirectory `python/pytorch`)? And then we could still have a notebook and maybe also a similar usage example in `python/examples`.

I have not looked at the code yet: did you test it to ensure the exploitability goes down in one of the benchmark games?

**Comment by MJ10** [comment]
Thanks for the response!

> I would love for this to be part of the main library, not just an example in a notebook. How do you feel about making it a callable module like all the other algorithms (maybe in a new subdirectory `python/pytorch`)? And then we could still have a notebook and maybe also a similar usage example in `python/examples`.

Yes sure. In the notebook, I have implemented a callable module itself. I wasn't sure if it should be part of the main library since an implementation of the algorithm already exists in the library. Will add the module to this PR itself. 
 
> I have not looked at the code yet: did you test it to ensure the exploitability goes down in one of the benchmark games?

Yes I have tested it on kuhn_poker. Should I test it on more games?

**Comment by lanctot** [comment]
Perfect, thanks. We have several copies of the same algorithm (e.g. in C++ and python) so it is ok to do this with TF vs Pytorch as well, and adding a pytorch directory now will make it easier for future PRs.

Yes I would love it of you could try on Leduc just to see what it gets to, since it is a very common benchmark. No need to make it a test (I guess it will take several hours) but we can compare to the value range we expect to find on Leduc.

Thanks again!

**Comment by lanctot** [comment]
Whoops, closed the PR by mistake, sorry!

**Comment by MJ10** [comment]
Thank you for the response!

> Perfect, thanks. We have several copies of the same algorithm (e.g. in C++ and python) so it is ok to do this with TF vs Pytorch as well, and adding a pytorch directory now will make it easier for future PRs.

Yes that makes sense, I will be pushing the module soon. 

> Yes I would love it of you could try on Leduc just to see what it gets to, since it is a very common benchmark. No need to make it a test (I guess it will take several hours) but we can compare to the value range we expect to find on Leduc.

I tried a small run on LeDuc with the following parameters: 
```python
      policy_network_layers=(32, 32),
      advantage_network_layers=(16, 16),
      num_iterations=400,
      num_traversals=40
```
With the TF Implementation: ```NashConv: 3.0775418037380993```
With the PyTorch Implementation: ```NashConv: 2.9058957077122525```

**Comment by lanctot** [comment]
Great! 400 iterations is not a lot.. how long did that take? Could you run it for 10000 or 10000, maybe computing the NashConv every 1000? 

You can just run your own if you like. I'm curious if we can get it to below 1.

**Comment by MJ10** [comment]
Yes sure, Thanks for the feedback! 
I have started runs with 10000 iterations. I will update here once done.

**Comment by MJ10** [comment]
Hey, so sorry for the long delay, got caught up with a few deadlines. 
So I ran the PyTorch and TF versions on Leduc, but was only able to reach NashConv of ~1.4 with both of them. 
Have also added the pytorch version to the library.

**Comment by lanctot** [comment]
Great, thanks @MJ10 ! Note that we made a few changes to our TF DeepCFR impl recently. 

1.4 NashConv hmm... after how many iterations? And what size nets did you use?

Would be great to sync with @aliyakamercan @ericsteinberger @noambrown on what good hyper-parameters are for Leduc. We don't need to block this PR to get the perfect setup, but would be good to have more confidence that we're reproducing results from the primary authors. 

Eric, Noam: did you ever run Deep CFR on Leduc with small networks (like, same-ish size as the ones used in the NFSP paper)? Would be great if you could point us to a convergence graph so we know what we're trying to achieve.

**Comment by MJ10** [comment]
@lanctot  Thanks! I didn't see the changes in the TF DeepCFR, will take a look at that!

File: open_spiel/python/pytorch/deep_cfr.py [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+
+
+AdvantageMemory = collections.namedtuple(
+    "AdvantageMemory", "info_state iteration advantage action")
+
+StrategyMemory = collections.namedtuple(
+    "StrategyMemory", "info_state iteration strategy_action_probs")
+
+
+def reset_params(m):
```
Comment (finbarrtimbers) [inline open_spiel/python/pytorch/deep_cfr.py:47]: Can you add a docstring here?

File: open_spiel/python/examples/deep_cfr_pytorch.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  
+  for player, losses in six.iteritems(advantage_losses):
+    logging.info("Advantage for player %d: %s", player,
+                  losses[:2] + ["..."] + losses[-2:])
+    logging.info("Advantage Buffer Size for player %s: '%s'", player,
+                  len(deep_cfr_solver.advantage_buffers[player]))
+  logging.info("Strategy Buffer Size: '%s'",
+                len(deep_cfr_solver.strategy_buffer))
+  logging.info("Final policy loss: '%s'", policy_loss)
+  conv = exploitability.nash_conv(
```
Comment (finbarrtimbers) [inline open_spiel/python/examples/deep_cfr_pytorch.py:65]: Can you use the C++ exploitability instead (pyspiel.nash_conv)? It's much faster.

File: open_spiel/python/pytorch/deep_cfr.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+
+    Over a traversal the advantage and strategy memories are populated with
+    computed advantage values and matched regrets respectively.
+
+    Args:
+      state: Current OpenSpiel game state.
+      player: (int) Player index for this traversal.
+
+    Returns:
+      Recursively returns expected payoffs for each action.
```
Comment (finbarrtimbers) [inline open_spiel/python/pytorch/deep_cfr.py:247]: Can you clarify what exactly is return? I believe it's a float for player.

File: open_spiel/python/pytorch/deep_cfr.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+        for a_ in state.legal_actions():
+          sampled_regret[action] -= strategy[a_] * expected_payoff[a_]
+        self._advantage_memories[player].add(
+            AdvantageMemory(state.information_state_tensor(),
+                            self._iteration, advantages, action))
+      return max(expected_payoff.values())
+    else:
+      other_player = state.current_player()
+      _, strategy = self._sample_action_from_advantage(state, other_player)
+      # Recompute distribution dor numerical errors.
```
Comment (finbarrtimbers) [inline open_spiel/python/pytorch/deep_cfr.py:275]: for

File: open_spiel/python/pytorch/deep_cfr.py [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+    matched_regrets = np.array([0.] * self._num_actions)
+    for action in legal_actions:
+      if cumulative_regret > 0.:
+        matched_regrets[action] = advantages[action] / cumulative_regret
+      else:
+        matched_regrets[action] = 1 / self._num_actions
+    return advantages, matched_regrets
+
+  def action_probabilities(self, state):
+    """Returns action probabilities dict for a single batch."""
```
Comment (finbarrtimbers) [inline open_spiel/python/pytorch/deep_cfr.py:310]: Can you provide a full docstring here?

File: open_spiel/python/pytorch/deep_cfr.py [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+    loss_advantages = self._loss_advantages(iters * outputs, iters * advantages)
+    loss_advantages.backward()
+    self._optimizer_advantages[player].step()
+    return loss_advantages.detach().numpy()
+
+  def _learn_strategy_network(self):
+    """Compute the loss over the strategy network.
+
+    Returns:
+      The average loss obtained on this batch of transitions or `None`.
```
Comment (finbarrtimbers) [inline open_spiel/python/pytorch/deep_cfr.py:357]: Can you mention what the type being returned is?

**Comment by MJ10** [comment]
Apologies for the extremely long delay. I got caught up with a few other things and didn't get a chance to push the changes

Thank you for the review @finbarrtimbers! I have made changes that you've requested. 

I have also made changes to match the recent fixes to the TF version!

**Comment by finbarrtimbers** [comment]
Hey sorry to miss this, looking now.

File: open_spiel/python/examples/deep_cfr_pytorch.py [hunk 7]
```diff
@@ -62,11 +61,19 @@ def main(unused_argv):
...
+
+  average_policy = policy.tabular_policy_from_callable(
+      game, deep_cfr_solver.action_probabilities)
+  pyspiel_policy = policy.python_policy_to_pyspiel_policy(average_policy)
+  conv = pyspiel.nash_conv(game, pyspiel_policy)
+  logging.info("Deep CFR in '%s' - NashConv: %s", FLAGS.game_name, conv)
+
+  average_policy_values = expected_game_score.policy_value(
+      game.new_initial_state(), [average_policy] * 2)
+  print("Computed player 0 value: {}".format(average_policy_values[0]))
```
Comment (finbarrtimbers) [inline open_spiel/python/examples/deep_cfr_pytorch.py:73]: Can you use logging.info here?

**Review by finbarrtimbers — CHANGES_REQUESTED** [review]
This lgtm once you fix the remaining print statements!

**Comment by MJ10** [comment]
Thanks @finbarrtimbers! I have fixed the print statements, I missed those earlier!

**Comment by finbarrtimbers** [comment]
Merging this internally- should be merged on Github early next week!

**Comment by MJ10** [comment]
Thank you @finbarrtimbers! :)

**Comment by finbarrtimbers** [comment]
Hey! Can you add your test to `python/CMakeLists.txt` & verify that it passes? You'll probably have to add additional imports to `requirements.txt`.

**Comment by lanctot** [comment]
Hi @MJ10 can you also take a look at https://github.com/deepmind/open_spiel/pull/406 ? I think you will need a similar change.

**Comment by MJ10** [comment]
This has been kept open for a while, apologies for the delay @lanctot @finbarrtimbers. I have included the test in the `CMakeLists.txt` and also included the fixes from #406 and #420

**Pull request merged**